Analítica de datos
Reto 2 semana 5. \
Carlos andrés Franco Gaitán

1. Obtener información sobre las variables.

En las especirficaciones del trabajo se pide que se tomen 8 variables de caracterización de los empleados, excluyendo la variable de personalidad, para aumentar el entendimiento propio y del espectador, vamos a describir cada una y su significado aquí.



*   Business: esta se entiende como la linea de negocios en la que está ubicada esta persona ya sea hardware que se representará con un 0, o software que se representará con un 1.

*   Age: esta es la edad que reporta la persona y se muestra en números discretos (numeros enteros sin decimales)

* Female: esta indica si la persona es de género femenino, el número 1 indica que la persona es una mujer, mientras que el 0 indica el caso contrario, osea que la persona es un hombre.

* Years: Esta es la cantidad de años que una persona ha trabajado continuamente para la compañía e igual que la edad se trabaja de con números enteros discretos.

* College: Si esta persona se ha graduado de educación universitaria de cuatro años, se representa 1 como si, y 0 como que no.

* Certificates: Este es el número de certificados de competencias profesionales que esta persona posee, esta tambien se expresa en números enteros.

* Feedback: la nota que ha recibido este empleado por parte de sus compañeros y supervisor de forma anual, la menor puntuación posible es cero (0) y la mayor es cuatro (4), esta escala va de 1 a 4 siendo 4 representante de un mejor rendimiento, y 1 de un rendimiento cuestionable, esta variable es numérica discreta, osea que si se encuentran decimales.


* Salary: este es el salario anual que gana cada empleado

* NPS: Este es el Net promoter score o Puntuación neta del promotor, este mide la lealtad del cliente. Según Qualtrics (2024), el NPS pregunta al cliente si recomendaría el producto, empresa o servicio a conocidos, el resultado de la respuesta de todos los clientes varía de -100 que significa que el 100% no recomendaría, a 100 que significa que el 100% recomendaría.
fuente: https://www.qualtrics.com/experience-management/customer/measure-nps/

Sin embargo, en este caso solo se tienen datos de 1 a 10 por lo que se va a asumir que en vez de una escala a 100 se utiliza una a 10, y que la ausencia de números negativos significa que ninguna medida NPS ha sido negativa, no que se mida de una forma distinta.






2. Cargar librerías

In [ ]:
#librerías de manejo de datos y operaciones matemáticas
import pandas as pd

import numpy as np

#libreria de implementación del arbol de decisión
from sklearn.tree import DecisionTreeClassifier

#Librería de evaluación del modelo
from sklearn.metrics import confusion_matrix

#Librería para crear el gráfico
from pydotplus import graph_from_dot_data #crear el grafico

#Librería para exportar el gráfico
from sklearn.tree import export_graphviz

#mostrar la imagen sin descargarla
from IPython.display import Image
import io

#cambiar info de letras a numeros
from sklearn.preprocessing import LabelEncoder


#crear conexión a google drive para no cargar la base de datos cada vez
#que cierre sesión
from google.colab import drive
#drive.mount('/content/drive')
#esto se utiliza mientras se desarrollaba el ejercicio, antes
#de entregarlo se cambia a metodo de carga a content


3. Cargar la base de datos

In [ ]:


bdatos = pd.read_excel("/content/2. TechSalesReport_Int_L.xlsx", sheet_name = 0 )

print(bdatos.head())


   Sales_Rep  Business  Age  Female  Years College Personality  Certificates  \
0          1  Hardware   59       1      2     Yes    Diplomat             1   
1          2  Hardware   52       0     10     Yes    Diplomat             4   
2          3  Software   47       1      1     Yes    Explorer             1   
3          4  Hardware   61       0      2     Yes    Diplomat             3   
4          5  Software   39       0      1      No    Diplomat             5   

   Feedback  Salary  NPS  
0      2.01   70200    5  
1      3.64  133000   10  
2      3.88   52600    8  
3      2.70   96000    6  
4      3.44  122000    7  


4. Transformar datos

In [ ]:


bdatos["College"] = bdatos["College"].map({"Yes": 1, "No": 0})

bdatos['Business'] = bdatos["Business"].map({"Hardware": 0, "Software": 1})
print(bdatos.head())

   Sales_Rep  Business  Age  Female  Years  College Personality  Certificates  \
0          1         0   59       1      2        1    Diplomat             1   
1          2         0   52       0     10        1    Diplomat             4   
2          3         1   47       1      1        1    Explorer             1   
3          4         0   61       0      2        1    Diplomat             3   
4          5         1   39       0      1        0    Diplomat             5   

   Feedback  Salary  NPS  
0      2.01   70200    5  
1      3.64  133000   10  
2      3.88   52600    8  
3      2.70   96000    6  
4      3.44  122000    7  


5. Verificar integridad de los datos

In [ ]:
print(pd.isna(bdatos).sum()) #no hay ningun NAN, no se hace nada mas

college = len(bdatos["College"])
college_true = len(bdatos[bdatos["College"] == 1])

print("Total de datos", college)

print("Gente con educacion universitaria", college_true)

print("el porcentaje de gente en la base de datos con titulo universitario es:",((college_true / college)*100))

Sales_Rep       0
Business        0
Age             0
Female          0
Years           0
College         0
Personality     0
Certificates    0
Feedback        0
Salary          0
NPS             0
dtype: int64
Total de datos 21990
Gente con educacion universitaria 17520
el porcentaje de gente en la base de datos con titulo universitario es: 79.67257844474761


Se encuentra un gran sesgo en la base de datos ya que mucho mas del 50% tienen educación universitaria, pero ante la falta de conocimiento de si este modelo es afectado como el de bayes si la info no es 50-50, o en caso que si como arreglarlo, no se hará nada al respecto.

6. Separar las variables

In [ ]:
xd = bdatos[["Business", "Age", "Female", "Years", "Certificates", "Feedback", "Salary", "NPS"]]
yd = bdatos["College"]

7. Implementación del modelo

In [ ]:
arbol = DecisionTreeClassifier(criterion = "gini", max_depth=4)
#se deciden usar 4 niveles debido a la recomendación del profe
#de usar un nivel por variable hasta llegar a un maximo de 4 especificados
#en el requerimiento del entregable.

arbol.fit(xd, yd)

DecisionTreeClassifier(max_depth=4)

In [ ]:
#Observarlo graficamente

variables = ["Business", "Age", "Female", "Years",
             "Certificates", "Feedback", "Salary", "NPS"]

dot_data = export_graphviz(arbol, feature_names = variables)

grafica = graph_from_dot_data(dot_data)
grafica.write_png("Arbol de decision ANEG reto S5.png")


True

8. Evaluar modelo

In [ ]:
ydp = arbol.predict(xd)
confusion = confusion_matrix(yd, ydp)
print(confusion)

VN = confusion[0,0]

FP = confusion[0,1]

FN = confusion[1,0]

VP = confusion[1,1]

Exactitud = (VP+VN)/(VP+VN+FP+FN)

Sensibilidad = VP/(VP+FN)

Especificidad = VN/(VN+FP)

Precisión = VP/(VP+FP)

print("Exactitud: ", Exactitud.round(3)*100, "%")
print("Sensibilidad: ", Sensibilidad.round(3)*100, "%")
print("Especificidad: ", Especificidad.round(3)*100, "%")
print("Precisión: ", Precisión*100, "%")


Tasa_de_error = (FP+FN)/(VP+VN+FP+FN)

print("Tasa de error: ", Tasa_de_error.round(3)*100, "%")

Prediccion_negativa = VN/(VN+FN)

print("Predicción negativa: ", Prediccion_negativa.round(3)*100, "%")


[[  114  4356]
 [   69 17451]]
Exactitud:  79.9 %
Sensibilidad:  99.6 %
Especificidad:  2.6 %
Precisión:  80.02476269087909 %
Tasa de error:  20.1 %
Predicción negativa:  62.3 %


9. Pronosticar con el modelo

In [ ]:
bdatos2 = pd.read_excel("/content/2. TechSalesReport_Int_L.xlsx", sheet_name = 1 )
#transformar
bdatos2['Business'] = bdatos2["Business"].map({"Hardware": 0, "Software": 1})
bdatos2.head()



Sales_Rep  Business  Age  Female  Years Personality  Certificates  \
0          1         0   63       0      2    Sentinel             5   
1          2         1   38       1      2    Diplomat             4   
2          3         1   47       0      1     Analyst             2   
3          4         1   25       0      1    Explorer             2   
4          5         0   49       0     10    Explorer             4   

   Feedback  Salary  NPS  
0      3.49  133000    8  
1      3.33   80400    6  
2      2.12   57000    7  
3      2.89   79000    7  
4      2.52  127000    6

In [ ]:
#Nota, asegurarse de correr la parte anterior antes de esta, ya que al final
#de esta se vuelve a convertir de números a nombres el hardware y software

#tomar variables
xd2 = bdatos2[["Business", "Age", "Female", "Years", "Certificates", "Feedback", "Salary", "NPS"]]

#prediccion
pronostico = arbol.predict(xd2)

print(pronostico)

bdatos2["Pronostico"] = pronostico
bdatos2["Business"] = bdatos2["Business"].map({0: "Hardware", 1: "Software"})


print(bdatos2)

bdatos2.to_excel("Pronostico reto 2.xlsx")


[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
    Sales_Rep  Business  Age  Female  Years Personality  Certificates  \
0           1  Hardware   63       0      2    Sentinel             5   
1           2  Software   38       1      2    Diplomat             4   
2           3  Software   47       0      1     Analyst             2   
3           4  Software   25       0      1    Explorer             2   
4           5  Hardware   49       0     10    Explorer             4   
5           6  Software   48       0      1    Diplomat             1   
6           7  Software   25       1      1    Explorer             6   
7           8  Hardware   39       0      1    Diplomat             2   
8           9  Software   29       0      1    Explorer             5   
9          10  Software   45       1      3    Diplomat             3   
10         11  Hardware   35       0      6    Diplomat             1   
11         12  Hardware   30       0      4    Diplomat             2   
12         13  So

10. Códigos varios para el posterior análisis

In [ ]:
print("La población es de", len(bdatos), "representantes de ventas")

es_mujer = len(bdatos[bdatos["Female"] == 1])

print("La cantidad de mujeres en la población es de", es_mujer,
      "y su porcentaje es de", es_mujer/len(bdatos)*100, "%")

print("la cantidad de hombres es de", len(bdatos)-es_mujer,
      "y su porcentaje es de", (len(bdatos)-es_mujer)/len(bdatos)*100, "%")

print("la persona mas joven tiene", min(bdatos["Age"]), "años")

print("la persona mas vieja tiene", max(bdatos["Age"]), "años")

print("el promedio de edad es de", bdatos["Age"].mean(), "años")


print("Gente con educacion universitaria", college_true)

print("el porcentaje de gente en la base de datos con \
titulo universitario es:",
 ((college_true / college)*100))

print("el salario mínimo es de", bdatos["Salary"].min(),
      "unidades monetarias")

print("el salario máximo es de", bdatos["Salary"].max())

print("el salario promedio es de", bdatos["Salary"].mean())

print("cantidad de personas que trabajan en software",
      len(bdatos[bdatos["Business"] == 1]))

print("cantidad de personas que trabajan en hardware",
      len(bdatos[bdatos["Business"] == 0]))

print("Porcentaje software",
      len(bdatos[bdatos["Business"] == 1])/len(bdatos)*100)

print("Porcentaje hardware",
      len(bdatos[bdatos["Business"] == 0])/len(bdatos)*100)

print("Mínimo de años en la compañía", min(bdatos["Years"]))

print("Máximo de años en la compañía", max(bdatos["Years"]))

print("Promedio de años en la compañía", bdatos["Years"].mean())

print("la media de años en la compañía", bdatos["Years"].median())

si = 1056
no = 45

print("el porcentaje de personas en este nodo sin titulo es de:",
      (no/(si+no))*100, "%")

print("el porcentaje de personas en este nodo con titulo es de:",
      (si/(si+no))*100, "%")

La población es de 21990 representantes de ventas
La cantidad de mujeres en la población es de 8423 y su porcentaje es de 38.30377444292861 %
la cantidad de hombres es de 13567 y su porcentaje es de 61.69622555707139 %
la persona mas joven tiene 21 años
la persona mas vieja tiene 65 años
el promedio de edad es de 41.49595270577535 años
Gente con educacion universitaria 17520
el porcentaje de gente en la base de datos con titulo universitario es: 79.67257844474761
el salario mínimo es de 21000 unidades monetarias
el salario máximo es de 197000
el salario promedio es de 73673.7789904502
cantidad de personas que trabajan en software 12130
cantidad de personas que trabajan en hardware 9860
Porcentaje software 55.16143701682583
Porcentaje hardware 44.83856298317417
Mínimo de años en la compañía 1
Máximo de años en la compañía 13
Promedio de años en la compañía 2.6463847203274216
la media de años en la compañía 2.0
el porcentaje de personas en este nodo sin titulo es de: 4.087193460490464 %


11. Análisis

En la base de datos suministrada se encuentran 21990 datos, cada uno conteniendo a un empleado de ventas, de estos 13567 son hombres (61.7%) y 8423 son mujeres (38.3%), entre estas personas la mas joven cuenta con 21 años, la mas longeva con 65, y el promedio de edad se encuentra en 41 años y medio, por lo que se puede observar que hay gran diversidad de personas en esta población.

De los casi 22000 empleados, casi el 80% (17520) cuentan con un titulo de educación universitaria, y en promedio ganan cerca de 74000 unidades monetarias (posiblemente dólares americanos), siendo el salario mínimo de 21000, y el máximo de casi 200 000, lo que demuestra una gran disparidad entre estos.

Sobre las areas de negocios podemos ver que la mayoría trabajan en el area de software (55%) frente a un casi 45% en hardware, siendo 1 la menor cantidad de años que algunos de estos llevan trabajando para la compañía, siendo el promedio un poco menor a 3, y el máximo de años 13. De estos números se podría interpretar dos situaciones, una que haya una gran rotación de personal, y otra que una gran expansión reciente haya llevado a la contratación masiva de nuevo personal causando una disminución en el promedio y la mediana.

Al evaluar el modelo, se halla una exactitud cercana al 80%, sensibilidad muy cercana al 100%, especificidad de casi el 3%, precisión del 80%, tasa de error del 20.1% y predicción negativa del 62.3%.

 De estos datos se puede concluir que el modelo es muy bueno pornosticando correctamente a los empleados que tienen titulo universitario, sin embargo la baja especificidad da a relucir que el modelo no es para nada competente identificando las personas que no tienen titulo universitario, por lo tanto se pueden esperar resultados con una cierta sobreestimación de la cantidad de personas con titulo universitario, pero otras metricas clave como la exactitud y la precisión dan muy buena cuenta de lo confiable que este modelo puede ser.

Sobre la racionalidad que hace el modelo en el arbol de decisión, encontramos que los valores gini mas relevantes se encuentran en la zona derecha inferior de este, y si bien no hay ningún nodo puro (gini = 0), si se encuentra un nodo muy cercano a ser puro cuya ruta es: salary > 64100 & certificates < 4.5 & Feedback < 3.495 en el cual hay 45 personas sin estudios universitarios (4%), y 1056 que sí (96%).

Entre los otros tres gini relevantes, uno compartía la variable feedback en el ultimo nivel (gini = 0.162), y otro utilizaba la variable salario > 109100 (gini = 0.138).

Utilizando este modelo se pronosticó que todas las personas de la base de datos para pronosticar, cuentan con estudios universitarios (ver informe excel adjunto).

